// json-joy/es6/reactive-rpc/server/http1/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WsConnectionContext = exports.Http1ConnectionContext = void 0;
const tslib_1 = require("tslib");
const util_1 = require("./util");
const concat_1 = require("../../../util/buffers/concat");
class Http1ConnectionContext {
    constructor(req, res, path, query, ip, token, params, meta, reqCodec, resCodec, msgCodec) {
        this.req = req;
        this.res = res;
        this.path = path;
        this.query = query;
        this.ip = ip;
        this.token = token;
        this.params = params;
        this.meta = meta;
        this.reqCodec = reqCodec;
        this.resCodec = resCodec;
        this.msgCodec = msgCodec;
    }
    body(maxPayload) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const list = yield (0, util_1.getBody)(this.req, maxPayload);
            const bodyUint8 = (0, concat_1.listToUint8)(list);
            return bodyUint8;
        });
    }
}
exports.Http1ConnectionContext = Http1ConnectionContext;
class WsConnectionContext {
    constructor(connection, path, query, ip, token, params, meta, reqCodec, resCodec, msgCodec) {
        this.connection = connection;
        this.path = path;
        this.query = query;
        this.ip = ip;
        this.token = token;
        this.params = params;
        this.meta = meta;
        this.reqCodec = reqCodec;
        this.resCodec = resCodec;
        this.msgCodec = msgCodec;
    }
}
exports.WsConnectionContext = WsConnectionContext;
