// json-joy/esm/server/services/PubSubService.js
import { Observable } from 'rxjs';
export class PubsubService {
    observers = new Map();
    listen$(channel) {
        return new Observable((observer) => {
            if (!this.observers.has(channel))
                this.observers.set(channel, []);
            const observers = this.observers.get(channel);
            observers.push(observer);
            return () => {
                const observers = this.observers.get(channel);
                if (!observers)
                    return;
                const index = observers.findIndex((o) => o === observer);
                if (index > -1)
                    observers.splice(index, 1);
                if (!observers.length) {
                    this.observers.delete(channel);
                }
            };
        });
    }
    async publish(channel, message) {
        await new Promise((resolve) => setImmediate(resolve));
        const observers = this.observers.get(channel);
        if (!observers)
            return;
        for (const observer of observers)
            observer.next(message);
    }
    stats() {
        return {
            channels: this.observers.size,
            observers: [...this.observers.values()].reduce((acc, v) => acc + v.length, 0),
        };
    }
}
