// json-joy/esm/reactive-rpc/common/rpc/caller/ObjectValueCaller.js
import { RpcError } from './error';
import { RpcCaller } from './RpcCaller';
import { FunctionStreamingType, FunctionType } from '../../../../json-type/type/classes';
import { StaticRpcMethod } from '../methods/StaticRpcMethod';
import { StreamingRpcMethod } from '../methods/StreamingRpcMethod';
export class ObjectValueCaller extends RpcCaller {
    router;
    system;
    methods = new Map();
    constructor({ router: value, ...rest }) {
        super({
            ...rest,
            getMethod: (name) => this.get(name),
        });
        this.router = value;
        const system = value.type.system;
        if (!system)
            throw new Error('NO_SYSTEM');
        this.system = system;
    }
    get(id) {
        let method = this.methods.get(id);
        if (method)
            return method;
        const fn = this.router.get(id);
        if (!fn || !(fn.type instanceof FunctionType || fn.type instanceof FunctionStreamingType)) {
            return undefined;
        }
        const fnType = fn.type;
        const { req, res } = fnType;
        const call = fn.data;
        const validator = fnType.req.validator('object');
        const requestSchema = fnType.req.getSchema();
        const isRequestVoid = requestSchema.__t === 'const' && requestSchema.value === undefined;
        const validate = isRequestVoid
            ? () => { }
            : (req) => {
                const error = validator(req);
                if (error) {
                    const message = error.message + (Array.isArray(error?.path) ? ' Path: /' + error.path.join('/') : '');
                    throw RpcError.value(RpcError.validation(message, error));
                }
            };
        method =
            fnType instanceof FunctionType
                ? new StaticRpcMethod({ req, res, validate, call })
                : new StreamingRpcMethod({ req, res, validate, call$: call });
        this.methods.set(id, method);
        return method;
    }
    async call(id, request, ctx) {
        return super.call(id, request, ctx);
    }
    async callSimple(id, request, ctx = {}) {
        try {
            const res = await this.call(id, request, ctx);
            return res.data;
        }
        catch (err) {
            const error = err;
            throw error.data;
        }
    }
    call$(id, request, ctx) {
        return super.call$(id, request, ctx);
    }
}
