// json-joy/es6/server/routes/util/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.util = exports.schema = exports.info = exports.echo = exports.ping = void 0;
const tslib_1 = require("tslib");
const ping = ({ t }) => (r) => {
    const Request = t.any;
    const Response = t.Const('pong');
    const Func = t.Function(Request, Response);
    return r.prop('util.ping', Func, () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        return 'pong';
    }));
};
exports.ping = ping;
const echo = ({ t }) => (r) => {
    const Request = t.any;
    const Response = t.any;
    const Func = t.Function(Request, Response);
    return r.prop('util.echo', Func, (msg) => tslib_1.__awaiter(void 0, void 0, void 0, function* () { return msg; }));
};
exports.echo = echo;
const info = ({ t, services }) => (r) => {
    const Request = t.any;
    const Response = t.Object(t.prop('now', t.num), t.prop('stats', t.Object(t.prop('pubsub', t.Object(t.prop('channels', t.num), t.prop('observers', t.num))), t.prop('presence', t.Object(t.prop('rooms', t.num), t.prop('entries', t.num), t.prop('observers', t.num))), t.prop('blocks', t.Object(t.prop('blocks', t.num), t.prop('patches', t.num))))));
    const Func = t.Function(Request, Response);
    return r.prop('util.info', Func, () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        return {
            now: Date.now(),
            stats: {
                pubsub: services.pubsub.stats(),
                presence: services.presence.stats(),
                blocks: services.blocks.stats(),
            },
        };
    }));
};
exports.info = info;
const schema = ({ t, router }) => (r) => {
    const Request = t.any;
    const Response = t.Object(t.prop('typescript', t.str));
    const Func = t.Function(Request, Response);
    return r.prop('util.schema', Func, () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        return {
            typescript: router.toTypeScript(),
        };
    }));
};
exports.schema = schema;
const util = (d) => (r) => ((0, exports.ping)(d)((0, exports.echo)(d)((0, exports.info)(d)((0, exports.schema)(d)(r)))));
exports.util = util;
