// json-joy/lib/json-ot/types/ot-string/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.type = exports.normalize = exports.transform = exports.compose = exports.apply = exports.validate = void 0;
const tslib_1 = require("tslib");
const validate_1 = require("./validate");
const util_1 = require("./util");
const apply_1 = require("./apply");
const compose_1 = require("./compose");
const transform_1 = require("./transform");
tslib_1.__exportStar(require("./types"), exports);
var validate_2 = require("./validate");
Object.defineProperty(exports, "validate", { enumerable: true, get: function () { return validate_2.validate; } });
var apply_2 = require("./apply");
Object.defineProperty(exports, "apply", { enumerable: true, get: function () { return apply_2.apply; } });
var compose_2 = require("./compose");
Object.defineProperty(exports, "compose", { enumerable: true, get: function () { return compose_2.compose; } });
var transform_2 = require("./transform");
Object.defineProperty(exports, "transform", { enumerable: true, get: function () { return transform_2.transform; } });
var util_2 = require("./util");
Object.defineProperty(exports, "normalize", { enumerable: true, get: function () { return util_2.normalize; } });
exports.type = {
    validate: validate_1.validate,
    normalize: util_1.normalize,
    apply: apply_1.apply,
    compose: compose_1.compose,
    transform: transform_1.transform,
};
