// json-joy/es6/server/services/PubSubService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PubsubService = void 0;
const tslib_1 = require("tslib");
const rxjs_1 = require("rxjs");
class PubsubService {
    constructor() {
        this.observers = new Map();
    }
    listen$(channel) {
        return new rxjs_1.Observable((observer) => {
            if (!this.observers.has(channel))
                this.observers.set(channel, []);
            const observers = this.observers.get(channel);
            observers.push(observer);
            return () => {
                const observers = this.observers.get(channel);
                if (!observers)
                    return;
                const index = observers.findIndex((o) => o === observer);
                if (index > -1)
                    observers.splice(index, 1);
                if (!observers.length) {
                    this.observers.delete(channel);
                }
            };
        });
    }
    publish(channel, message) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield new Promise((resolve) => setImmediate(resolve));
            const observers = this.observers.get(channel);
            if (!observers)
                return;
            for (const observer of observers)
                observer.next(message);
        });
    }
    stats() {
        return {
            channels: this.observers.size,
            observers: [...this.observers.values()].reduce((acc, v) => acc + v.length, 0),
        };
    }
}
exports.PubsubService = PubsubService;
