// json-joy/esm/reactive-rpc/server/ws/codec/WsFrameDecoder.js
import { StreamingOctetReader } from '../../../../util/buffers/StreamingOctetReader';
import { WsFrameDecodingError } from './errors';
import { WsCloseFrame, WsFrameHeader, WsPingFrame, WsPongFrame } from './frames';
export class WsFrameDecoder {
    reader = new StreamingOctetReader();
    push(uint8) {
        this.reader.push(uint8);
    }
    readFrameHeader() {
        try {
            const reader = this.reader;
            if (reader.size() < 2)
                return undefined;
            const b0 = reader.u8();
            const b1 = reader.u8();
            const fin = (b0 >>> 7);
            const opcode = b0 & 0b1111;
            const maskBit = b1 >>> 7;
            let length = b1 & 0b01111111;
            if (length === 126) {
                if (reader.size() < 2)
                    return undefined;
                length = (reader.u8() << 8) | reader.u8();
            }
            else if (length === 127) {
                if (reader.size() < 8)
                    return undefined;
                reader.skip(4);
                length = reader.u32();
            }
            let mask;
            if (maskBit) {
                if (reader.size() < 4)
                    return undefined;
                mask = [reader.u8(), reader.u8(), reader.u8(), reader.u8()];
            }
            if (opcode >= 8) {
                switch (opcode) {
                    case 8: {
                        return new WsCloseFrame(fin, opcode, length, mask, 0, '');
                    }
                    case 9: {
                        if (length > 125)
                            throw new WsFrameDecodingError();
                        const data = mask ? reader.bufXor(length, mask, 0) : reader.buf(length);
                        return new WsPingFrame(fin, opcode, length, mask, data);
                    }
                    case 10: {
                        if (length > 125)
                            throw new WsFrameDecodingError();
                        const data = mask ? reader.bufXor(length, mask, 0) : reader.buf(length);
                        return new WsPongFrame(fin, opcode, length, mask, data);
                    }
                    default: {
                        throw new WsFrameDecodingError();
                    }
                }
            }
            return new WsFrameHeader(fin, opcode, length, mask);
        }
        catch (err) {
            if (err instanceof RangeError)
                return undefined;
            throw err;
        }
    }
    readFrameData(frame, remaining, dst, pos) {
        const reader = this.reader;
        const mask = frame.mask;
        const readSize = Math.min(reader.size(), remaining);
        if (!mask)
            reader.copy(readSize, dst, pos);
        else {
            const alreadyRead = frame.length - remaining;
            reader.copyXor(readSize, dst, pos, mask, alreadyRead);
        }
        return remaining - readSize;
    }
    copyFrameData(frame, dst, pos) {
        const reader = this.reader;
        const mask = frame.mask;
        const readSize = frame.length;
        if (!mask)
            reader.copy(readSize, dst, pos);
        else
            reader.copyXor(readSize, dst, pos, mask, 0);
    }
    readCloseFrameData(frame) {
        let length = frame.length;
        if (length > 125)
            throw new WsFrameDecodingError();
        let code = 0;
        let reason = '';
        if (length > 0) {
            if (length < 2)
                throw new WsFrameDecodingError();
            const reader = this.reader;
            const mask = frame.mask;
            const octet1 = reader.u8() ^ (mask ? mask[0] : 0);
            const octet2 = reader.u8() ^ (mask ? mask[1] : 0);
            code = (octet1 << 8) | octet2;
            length -= 2;
            if (length)
                reason = reader.utf8(length, mask ?? [0, 0, 0, 0], 2);
        }
        frame.code = code;
        frame.reason = reason;
    }
}
