// json-joy/es6/json-patch/op/Op.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpExtend = exports.OpMerge = exports.OpSplit = exports.OpStrDel = exports.OpStrIns = exports.OpInc = exports.OpFlip = exports.OpCopy = exports.OpMove = exports.OpReplace = exports.OpRemove = exports.OpAdd = exports.OpType = exports.OpMatches = exports.OpNot = exports.OpOr = exports.OpAnd = exports.OpMore = exports.OpLess = exports.OpIn = exports.OpStarts = exports.OpEnds = exports.OpContains = exports.OpTestStringLen = exports.OpTestString = exports.OpTestType = exports.OpUndefined = exports.OpDefined = exports.OpTest = exports.AbstractPredicateOp = void 0;
const OpAdd_1 = require("./OpAdd");
Object.defineProperty(exports, "OpAdd", { enumerable: true, get: function () { return OpAdd_1.OpAdd; } });
const OpRemove_1 = require("./OpRemove");
Object.defineProperty(exports, "OpRemove", { enumerable: true, get: function () { return OpRemove_1.OpRemove; } });
const OpReplace_1 = require("./OpReplace");
Object.defineProperty(exports, "OpReplace", { enumerable: true, get: function () { return OpReplace_1.OpReplace; } });
const OpMove_1 = require("./OpMove");
Object.defineProperty(exports, "OpMove", { enumerable: true, get: function () { return OpMove_1.OpMove; } });
const OpCopy_1 = require("./OpCopy");
Object.defineProperty(exports, "OpCopy", { enumerable: true, get: function () { return OpCopy_1.OpCopy; } });
const OpTest_1 = require("./OpTest");
Object.defineProperty(exports, "OpTest", { enumerable: true, get: function () { return OpTest_1.OpTest; } });
const OpFlip_1 = require("./OpFlip");
Object.defineProperty(exports, "OpFlip", { enumerable: true, get: function () { return OpFlip_1.OpFlip; } });
const OpInc_1 = require("./OpInc");
Object.defineProperty(exports, "OpInc", { enumerable: true, get: function () { return OpInc_1.OpInc; } });
const OpStrIns_1 = require("./OpStrIns");
Object.defineProperty(exports, "OpStrIns", { enumerable: true, get: function () { return OpStrIns_1.OpStrIns; } });
const OpStrDel_1 = require("./OpStrDel");
Object.defineProperty(exports, "OpStrDel", { enumerable: true, get: function () { return OpStrDel_1.OpStrDel; } });
const OpSplit_1 = require("./OpSplit");
Object.defineProperty(exports, "OpSplit", { enumerable: true, get: function () { return OpSplit_1.OpSplit; } });
const OpMerge_1 = require("./OpMerge");
Object.defineProperty(exports, "OpMerge", { enumerable: true, get: function () { return OpMerge_1.OpMerge; } });
const OpExtend_1 = require("./OpExtend");
Object.defineProperty(exports, "OpExtend", { enumerable: true, get: function () { return OpExtend_1.OpExtend; } });
const OpDefined_1 = require("./OpDefined");
Object.defineProperty(exports, "OpDefined", { enumerable: true, get: function () { return OpDefined_1.OpDefined; } });
const OpUndefined_1 = require("./OpUndefined");
Object.defineProperty(exports, "OpUndefined", { enumerable: true, get: function () { return OpUndefined_1.OpUndefined; } });
const OpTestType_1 = require("./OpTestType");
Object.defineProperty(exports, "OpTestType", { enumerable: true, get: function () { return OpTestType_1.OpTestType; } });
const OpTestString_1 = require("./OpTestString");
Object.defineProperty(exports, "OpTestString", { enumerable: true, get: function () { return OpTestString_1.OpTestString; } });
const OpTestStringLen_1 = require("./OpTestStringLen");
Object.defineProperty(exports, "OpTestStringLen", { enumerable: true, get: function () { return OpTestStringLen_1.OpTestStringLen; } });
const OpContains_1 = require("./OpContains");
Object.defineProperty(exports, "OpContains", { enumerable: true, get: function () { return OpContains_1.OpContains; } });
const OpEnds_1 = require("./OpEnds");
Object.defineProperty(exports, "OpEnds", { enumerable: true, get: function () { return OpEnds_1.OpEnds; } });
const OpStarts_1 = require("./OpStarts");
Object.defineProperty(exports, "OpStarts", { enumerable: true, get: function () { return OpStarts_1.OpStarts; } });
const OpIn_1 = require("./OpIn");
Object.defineProperty(exports, "OpIn", { enumerable: true, get: function () { return OpIn_1.OpIn; } });
const OpLess_1 = require("./OpLess");
Object.defineProperty(exports, "OpLess", { enumerable: true, get: function () { return OpLess_1.OpLess; } });
const OpMore_1 = require("./OpMore");
Object.defineProperty(exports, "OpMore", { enumerable: true, get: function () { return OpMore_1.OpMore; } });
const OpAnd_1 = require("./OpAnd");
Object.defineProperty(exports, "OpAnd", { enumerable: true, get: function () { return OpAnd_1.OpAnd; } });
const OpOr_1 = require("./OpOr");
Object.defineProperty(exports, "OpOr", { enumerable: true, get: function () { return OpOr_1.OpOr; } });
const OpNot_1 = require("./OpNot");
Object.defineProperty(exports, "OpNot", { enumerable: true, get: function () { return OpNot_1.OpNot; } });
const OpMatches_1 = require("./OpMatches");
Object.defineProperty(exports, "OpMatches", { enumerable: true, get: function () { return OpMatches_1.OpMatches; } });
const OpType_1 = require("./OpType");
Object.defineProperty(exports, "OpType", { enumerable: true, get: function () { return OpType_1.OpType; } });
var AbstractPredicateOp_1 = require("./AbstractPredicateOp");
Object.defineProperty(exports, "AbstractPredicateOp", { enumerable: true, get: function () { return AbstractPredicateOp_1.AbstractPredicateOp; } });
