// json-joy/es6/json-patch/op/OpReplace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpReplace = void 0;
const AbstractOp_1 = require("./AbstractOp");
const json_pointer_1 = require("../../json-pointer");
class OpReplace extends AbstractOp_1.AbstractOp {
    constructor(path, value, oldValue) {
        super(path);
        this.value = value;
        this.oldValue = oldValue;
    }
    op() {
        return 'replace';
    }
    code() {
        return 2;
    }
    apply(doc) {
        const ref = (0, json_pointer_1.find)(doc, this.path);
        if (ref.val === undefined)
            throw new Error('NOT_FOUND');
        if ((0, json_pointer_1.isObjectReference)(ref))
            ref.obj[ref.key] = this.value;
        else if ((0, json_pointer_1.isArrayReference)(ref))
            ref.obj[ref.key] = this.value;
        else
            doc = this.value;
        return { doc, old: ref.val };
    }
    toJson(parent) {
        const json = {
            op: 'replace',
            path: (0, json_pointer_1.formatJsonPointer)(this.path),
            value: this.value,
        };
        if (this.oldValue !== undefined)
            json.oldValue = this.oldValue;
        return json;
    }
    toCompact(parent, verbose) {
        const opcode = verbose ? 'replace' : 2;
        return this.oldValue === undefined
            ? [opcode, this.path, this.value]
            : [opcode, this.path, this.value, this.oldValue];
    }
    encode(encoder, parent) {
        const hasOldValue = this.oldValue !== undefined;
        encoder.encodeArrayHeader(hasOldValue ? 4 : 3);
        encoder.writer.u8(2);
        encoder.encodeArray(this.path);
        encoder.encodeAny(this.value);
        if (hasOldValue)
            encoder.encodeAny(this.oldValue);
    }
}
exports.OpReplace = OpReplace;
