// json-joy/es2020/json-patch/op/OpNot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpNot = void 0;
const AbstractSecondOrderPredicateOp_1 = require("./AbstractSecondOrderPredicateOp");
const json_pointer_1 = require("../../json-pointer");
class OpNot extends AbstractSecondOrderPredicateOp_1.AbstractSecondOrderPredicateOp {
    constructor(path, ops) {
        super(path, ops);
        this.ops = ops;
    }
    op() {
        return 'not';
    }
    code() {
        return 44;
    }
    test(doc) {
        for (const op of this.ops)
            if (op.test(doc))
                return false;
        return true;
    }
    toJson(parent) {
        const op = {
            op: 'not',
            path: (0, json_pointer_1.formatJsonPointer)(parent ? this.path.slice(parent.path.length) : this.path),
            apply: this.ops.map((op) => op.toJson(this)),
        };
        return op;
    }
    toCompact(parent, verbose) {
        const opcode = verbose ? 'not' : 44;
        return [
            opcode,
            parent ? this.path.slice(parent.path.length) : this.path,
            this.ops.map((op) => op.toCompact(this, verbose)),
        ];
    }
    encode(encoder, parent) {
        encoder.encodeArrayHeader(3);
        encoder.writer.u8(44);
        encoder.encodeArray(parent ? this.path.slice(parent.path.length) : this.path);
        const length = this.ops.length;
        encoder.encodeArrayHeader(length);
        for (let i = 0; i < length; i++)
            this.ops[i].encode(encoder, this);
    }
}
exports.OpNot = OpNot;
